src/notebooks/execution/cellExecution.ts

```diff
@@ -503,26 +355,22 @@ export class CellExecution implements IDisposable {
             traceError(`Cell execution failed without request, for cell Index ${this.cell.index}`, ex);
             return this.completedWithErrors(ex);
         }
-        // Listen to messages and update our cell execution state appropriately
-        // Keep track of our clear state
-        const clearState = new RefBool(false);
-
-        const request = this.request;
-        request.onIOPub = (msg) => {
-            // Cell has been deleted or the like.
-            if (this.cell.document.isClosed) {
-                request.dispose();
-            }
-            this.handleIOPub(clearState, msg);
-        };
-        request.onReply = (msg) => {
-            // Cell has been deleted or the like.
-            if (this.cell.document.isClosed) {
-                request.dispose();
-            }
-            this.handleReply(clearState, msg);
-        };
-        request.onStdin = this.handleInputRequest.bind(this, session);
+        this.cellExecutionHandler = this.factory.create(this.cell, {
```

Comment:
I'm wondering if the 'cellExecutionHandler' could just have a single method on it, 'dispose()' that's called when ending. 

Otherwise you could push the onErrorHandlingExecuteRequestIOPubMessage into the factory.create (don't need to sign up for it separately).

Then the factory would be more like: registerListener or something that returns a disposable. Essentially it's listening to the request.

Reply:
Then all you need to do in 'endCellTask' (the only spot that uses cellExecutionHandler)  is stop listening - dispose the listener.

Reply:
Done, 
* Left dipose method so that this is only disposed when the cellExecution dies (called durijng dispose of cellExecution)
* endCellExecution is called internally by this class when `request.done` completes (either rejects or resolves)
* that way calling code (CellExecution) doesn't need to call anything)

Reply:
Basically `endCellExecution` is no longer a public method, & as suggested we only call dispose.

Reply:
What do you think of changing it to registerListener or maybe watchRequest? Instead of a factory.create? It's more like you're registering something to listen/watch a request? The semantics of 'create' seem strange to me.

Reply:
Fixed